### ISSUE TITLE
Create Component Manager

### ISSUE THREAD

**Issue by smudge202** [clean-development/engine#5]
I've mentioned ECS a couple times and we're currently working through a not so pleasant implementation in the [smudge202/storybox](https://github.com/smudge202/storybox) advanced stream.

Essentially, what we need is the abstractions and at least one implementation that does the complies with the following:

_typed in browser so forgive typos_

``` c#

[Unit]
public static void WhenEntityIsNullThenThrowsException()
{
  Action act = () => CreateTarget().Assign<object>(null);
  act.ShouldThrow<ArgumentNullException>();
}

[Unit]
public static void WhenAssigningNullInstanceToEntityThenThrowsException()
{
  Action act = () => CreateTarget().Assign(new object(), null);
  act.ShouldThrow<ArgumentNullException>();
}

[Unit]
public static void WhenInstanceAssignedThenRetrievesSameInstance()
{
  var trait = new object();
  var entity = new object();
  var target = CreateTarget();

  target.Assign(entity, trait);

  target.Retieve<object>(entity)
    .Should().Be(trait);
}

[Unit]
public static void WhenAssigningMultipleInstancesOfSameTypeOfTraitToSameEntityThenThrowsException()
{
  var entity = new object();
  var target = CreateTarget();
  target.Assign(entity, new object());
  Action act = () => target.Assign(entity, new object());

  act.ShouldThrow<InvalidOperationException>();
}

[Unit]
public static void WhenRetrievingTraitNotAssignedToEntityThenReturnsNull()
{
  CreateTarget().Retieve<object>(new object())
    .Should().BeNull();
}

[Unit]
public static void WhenRetrievingTraitAssignedToDifferentEnttityThenReturnsNull()
{
  var target = CreateTarget();
  target.Assign<object>(new object());

  target.Retieve<object>(new object())
    .Should().BeNull();
}

[Unit]
public static void WhenRetrievedTraitUpdatedThenSameTraitRetrievedContainsUpdate()
{
  var target = CreateTarget();
  var entity = new object();

  target.Assign<Fake.Trait>(entity);
  var trait = target.Retrieve<Fake.Trait>(entity);
  var value  = RandomString;
  trait.Property = value;

  target.Retrieve<Fake.Trait>(entity).Property
    .Should().Be(value);
}

[Unit]
public static void WhenRetrievingUnassignedTraitThenReturnsNull()
{
  var target  = CreateTarget();
  var entity = new object();

  target.Assign<object>();
  target.UnAssign<object>();

  target.Retrieve<object>()
    .Should().BeNull();
}
```

There's probably a lot more, but that's hopefully enough to get someone started?

**Comment by smudge202** [comment]
Blocked by #2

**Comment by danielcirket** [comment]
I might give this a go if no-one else wants to pick it up once #2 is done.

**Comment by smudge202** [comment]
Updated 20-10-2015:

~~Having slept on it, I don't think #9 resolves this (even at MVP level).~~

~~For starts, we need the ability to unassign a Component. That shouldn't be a big deal to add.
I also want to remove the `Assign` method and rename the `TryAssign` to `Assign`, which will allow me to drop the `ComponentAlreadyAssignedException`.~~

_^ Implemented in recent updates to pr #9 ._

Later (possibly post MVP, but at least under another story) I want to add some Rx/Observable functionality around Component assignment/removal, as well as the ability to invert the querying mechanism so that all entities with a given component assigned can be returned. To improve the latter, I may add some Generic based type filtering mechanisms so that components can filter down the entities that are allowed to have the component assigned. I'll get this last part all into a new issue once I've given it some more thought and come up with some use cases.
